feat: refactor multicurve

ieow opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Jira Link:
In this pr, the mpc corekit will be keyTypes and sigType agnostic.
it will throw error only during signing if the required keyTypes and tssLib is not available.

all keyTypes for
- getTssData will have same
   - tssFactorPubs
- getTssShare for same factor key
   - tssShareIndex  


check the added new tests on how it can be call for different signing scheme
## Description
<!--- Describe your changes in detail -->
Refactor multicurve 

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project. (run lint)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My code requires a db migration.

ieow committed:
feat: initial commit
integrate multicurve tkey
tests passed

ieow committed:
fix: add randomId verifier testcases

ieow committed:
fix: enable multicurve
fix tests
add multicurve tests

ieow committed:
feat: refactor multicurve

ieow committed:
fix: allow tsslib to be added later

ieow committed:
fix: multicurve tests

ieow committed:
fix: stable with most test passed

ieow committed:
fix: import recovery test

ieow committed:
feat: refactor setTssTag

matthiasgeihs commented on src/interfaces.ts:
```diff
@@ -487,14 +485,14 @@ export interface ICoreKit {
+  //     secp256k1Precompute?: Secp256k1PrecomputedClient;
+  //     keyTweak?: BN;
+  //   }
+  // ): Promise<Buffer>;
```
remove if not needed.

why does the core kit interface no longer have a `sign` method? isn't this the most important method?

ieow replied:
update with relevant function signatures and comments

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -228,40 +208,61 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
 
-  private get useClientGeneratedTSSKey(): boolean {
-    return this._sigType === "ed25519" && this.options.useClientGeneratedTSSKey === undefined ? true : !!this.options.useClientGeneratedTSSKey;
+  // private get useClientGeneratedTSSKey(): boolean {
+  //   return this._sigType === "ed25519" && this.options.useClientGeneratedTSSKey === undefined ? true : !!this.options.useClientGeneratedTSSKey;
+  // }
```
remove?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -228,40 +208,61 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
 
-  public getTssData(args: { skipThrow: boolean; keyType?: KeyType } = { skipThrow: false }) {
-    const result = this.tkey.metadata.getTssData(args.keyType ?? this.keyType, TSS_TAG_DEFAULT);
+  public getTssData(args: { skipThrow?: boolean; keyType: KeyType }) {
```
this `skipThrow` flag looks super odd. if you want to skip an error, just catch and discard!?

ieow replied:
break the function to 2 functions for more clearity

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -292,30 +293,36 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
     return finalKey.toString("hex", 64);
   }
 
+  public getServiceProviderKeyType() {
```
This does not access the service provider. Why is it called `getServiceProviderKeyType`?
Add function comment to explain further.

matthiasgeihs replied:
Why is this a public function?

ieow replied:
It is used to decide what service provider's keyType that mpc corekit should instantiate during init.

in multicurve, it is defaulted to secp256k1 only
in legacy, it is depend on the keyType that is configured 

will move to private

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -292,30 +293,36 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
-    if (this._sigType === "ed25519" && this.options.useDKG) {
-      throw CoreKitError.invalidConfig("DKG is not supported for ed25519 signature type");
-    }
+    const nodeDetails = fetchLocalConfig(this.options.web3AuthNetwork, KeyType.ed25519);
```
Why do you fix the key type to `ed25519`?

ieow replied:
the nodedetails variables is used to get the metadata url.
It does not matter which keyType is used to get the node details.
We can replace with the spKeyType for consistency

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -476,11 +483,12 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
       throw CoreKitError.invalidConfig("Cannot import TSS key and register SFA key at the same time.");
     }
 
+    const fisrtKeyType = this.getSupportedCurveKeyTypes()[0];
```
```suggestion
    const firstKeyType = this.getSupportedCurveKeyTypes()[0];
```

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -476,11 +483,12 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
       const prefetchTssPubs = [];
       for (let i = 0; i < prefetchTssPublicKeys; i++) {
-        prefetchTssPubs.push(this.torusSp.getTSSPubKey(this.tkey.tssTag, i, this.keyType));
+        prefetchTssPubs.push(this.torusSp.getTSSPubKey(this.tkey.tssTag, i, fisrtKeyType));
```
why do you use the first key type here? where is this documented?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -581,8 +594,9 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
+      const spKeyType = this.getServiceProviderKeyType();
       // Check for existing curve in tssData for legacy mode
-      const tssData = this.getTssData({ skipThrow: true });
+      const tssData = this.getTssData({ keyType: spKeyType, skipThrow: true });
```
legacy mode key type always equals service provider key type?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -759,12 +774,13 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
-    const tssCurve = getKeyCurve(this.keyType);
-    const p = this.getPubKeyPoint().toEllipticPoint(tssCurve);
+    const tssCurve = getKeyCurve(KeyType.ed25519);
+    // ed25519 only support account index 0 ?
+    const p = this.getPubKeyPoint(KeyType.ed25519, 0).toEllipticPoint(tssCurve);
```
```suggestion
    const p = this.getPubKeyPoint(KeyType.ed25519).toEllipticPoint(tssCurve);
```
i don't think it's a good idea to just assume index 0 here.
better throw an error if the user is on unsupported account index. alternatively, simply return the result.

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -872,39 +891,66 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
-      hashed?: boolean;
-      secp256k1Precompute?: Secp256k1PrecomputedClient;
-      keyTweak?: BN;
+  public async signECDSA(data: Uint8Array, opts?: { hashed?: boolean; secp256k1Precompute?: Secp256k1PrecomputedClient }) {
```
This is confusing. Precompute function is called `precompute_secp256k1`. Sign function is called `signECDSA`. Should be consistent!?
Moreover, ECDSA could be any curve. Do you always mean ECDSA over curve secp256k1 here?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -872,39 +891,66 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
+    const { hashed = false, secp256k1Precompute } = opts || {};
 
-      return this.sign_frost(data, opts?.keyTweak);
+    // TODO: replace buffer to uint8array
```
address TODO and remove comment

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -872,39 +891,66 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
+    return Buffer.concat([sig.r, sig.s, Buffer.from([sig.v])]);
+  }
+
+  public async signBip340(data: Uint8Array, opts?: { hashed?: boolean; keyTweak?: BN }) {
```
i'm still torn on whether it is better to keep the `sign` function with a `keyType` argument instead? what were the arguments for doing it with two separate functions like here?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -872,39 +891,66 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
       throw CoreKitError.factorKeyNotPresent("factorKey not present in state when deleting a factor.");
     }
-    const tssData = this.getTssData();
+    const firstKeyType = this.getSupportedCurveKeyTypes()[0];
```
why is first key type correct here?

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1043,7 +1084,8 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
    * For signature type ed25519, consider using _UNSAFE_exportTssEd25519Seed.
    */
-  public async _UNSAFE_exportTssKey(): Promise<string> {
+  // TODO : should return all keyType final keys?
```
address TODO and remove

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1128,15 +1170,19 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
-          importKey.secp256k1 = scalarBNToBufferSEC1(k).toString("hex");
-        } else {
-          throw CoreKitError.default(`Unsupported key type and sig type combination: ${this.keyType}, ${this._sigType}`);
+      // if (!importKey && this.useClientGeneratedTSSKey) {
```
remove

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1185,8 +1231,8 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
   //   );
   // }
```
remove

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1210,34 +1256,42 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
-      }
-      if (this.supportedCurveKeyTypes.has(KeyType.ed25519)) {
+
+        if (this.supportedCurveKeyTypes.has(KeyType.ed25519)) {
+          const importTssBufEd25519 = importTssKey.ed25519 ? Buffer.from(importTssKey.ed25519, "hex") : undefined;
+          // check if key is in the tsslib and keytype exists
+          await this.tKey.initializeTss({
+            importKey: importTssBufEd25519,
+            tssKeyType: KeyType.ed25519,
+            serverOpts: {
+              // selectedServers: [],
+              authSignatures: this.state.signatures,
+            },
+          });
+        }
+      } else {
```
hu? why do you duplicate this part from below and nest it here? just move the below part out of the else and condition it like here!?

ieow replied:
it is different for initialising first curve and 2nd curve.
during initializing 2nd curve, factorPub, and tssIndex are not allowed to be provided,

updating the logic and comment for easier reading

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1465,7 +1513,8 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
   private async copyOrCreateShare(newFactorTSSIndex: number, newFactorPub: Point) {
     this.checkReady();
-    const tssData = this.getTssData();
+    const firstKeyType = this.getSupportedCurveKeyTypes()[0];
```
Since you use this a lot, I think you should create a dedicated function for it.
What's the logic behind always using the first key type here?

ieow replied:
I are trying to get tssIndex from getTssShare or factorPub from getTssData.
both curves should have the same tssIndex (for same factorKey) and factorPub
Using either curve's keyType will get the same value.

using the first as instance might configured with single curve keytype only

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1543,7 +1592,8 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
 
-    const { tssIndex } = await this.getTssShare(factorKey);
+    const firstKeyType = this.getSupportedCurveKeyTypes()[0];
+    const { tssIndex } = await this.getTssShare({ keyType: firstKeyType, factorkey: factorKey });
```
Isn't this wrong? What if you want to get the `ed25519` share but first key type is `secp256k1`? Same applies to several other instances of first key type usage.

ieow replied:
I are trying to get tssIndex from getTssShare or factorPub from getTssData.
both curves should have the same tssIndex (for same factorKey) and factorPub 
Using either curve's keyType will get the same value.

using the first as instance might configured with single curve keytype only

matthiasgeihs commented on src/mpcCoreKit.ts:
```diff
@@ -1642,8 +1692,15 @@ export class Web3AuthMPCCoreKit implements ICoreKit, IMPCContext {
+  private async sign_frost(args: {
+    data: Buffer;
+    keyTweak?: BN;
+    keyType: KeyType;
```
you can derive the key type from the sig type

ieow replied:
updated

matthiasgeihs commented on src/utils.ts:
```diff
@@ -211,5 +211,37 @@ export function makeEthereumSigner(kit: CoreKitSigner): EthereumSigner {
   };
 }
 
+// export function makeBip340Signer(kit: CoreKitSigner): EthereumSigner {
+//   if (kit.keyType !== KeyType.secp256k1) {
+//     throw new Error(`Invalid key type: expected secp256k1, got ${kit.keyType}`);
+//   }
+//   return {
+//     sign: async (msgHash: Buffer) => {
+//       const sig = await kit.sign(msgHash, { hashed: true });
+//       return sigToRSV(sig);
+//     },
+//     getPublic: async () => {
+//       const pk = Point.fromSEC1(secp256k1, kit.getPubKey().toString("hex"));
+//       return pk.toSEC1(secp256k1).subarray(1);
+//     },
+//   };
+// }
```
remove if not needed

ieow replied:
Initially thought the makeSigner for other is not required.
Added back now to make developer easier to understand and integrate.

matthiasgeihs commented:
i think it would be good to merge this into the multi curve branch and then review together. why are the two PRs separated? @ieow

ieow committed:
Merge pull request #218 from Web3Auth/feat/refactor-v2-multicurve-setTssTag

feat: refactor setTssTag

ieow committed:
fix: comments
remove unsed code and comment
update new user comments and flow
update comment

ieow committed:
fix: get keyType and SigType from frostLib

ieow committed:
fix: update interface and comment
rename function name

ieow committed:
feat: break getTssData to 2 function for clearity

ieow commented:
> i think it would be good to merge this into the multi curve branch and then review together. why are the two PRs separated? @ieow

The first pr is initial solution using setTkeyType to set active keyType
this current pr is keyType agnostic ( no active keyType)
previously there was was on discussion to choose which direction.
It was decided to use the agnostic direction.
Previous PR should be closed. I closed it yesterday.

Recent added new pr is support tkey's pr which add back the setTssTag instead of passing tag to most of the tkey tss function

since the prs is merged in tkey side, mpc-corekit prs is also merged 

now there is only single pr in mpc core kit and tkey tss

ieow committed:
fix: address comment
fix ed25519 with accountIndex 0
fix useClientGeneratedTSSKey
fix typo

ieow committed:
fix: rename signECDSA function name
fix multicurve test to use new storage instance
fix useClientGeneratedTSSKey bug

ieow committed:
fix: update demo app
